Kafka consumer r50 (#1788)
emqtt/emq-docs-cn: en_US/data-integration/data-bridge-kafka.md

# Stream Data into Apache Kafka

<!-- 提供一段简介，描述支数据桥接的基本工作方式、关键特性和价值，如果有局限性也应当在此处说明（如必须说明的版本限制、当前未解决的问题）。 -->

> [Apache Kafka](https://kafka.apache.org/) is a widely-used open-source distributed event streaming platform. EMQX's integration with Apache Kafka/Confluent presents our users with reliable bi-directional data transport and processing capability under high-throughput scenarios. As a top IoT data infrastructure provider, EMQX currently supports authenticating with Apache Kafka/Confluent via SASL/SCRAM or SASL/GSSAPI. 

Reviewer: ```suggestion
[Apache Kafka](https://kafka.apache.org/) is a popular open-source distributed event streaming platform. EMQX's integration with Apache Kafka/Confluent offers users dependable bi-directional data transport and processing capabilities in high-throughput situations. Furthermore, EMQX currently supports authentication with Apache Kafka/Confluent using SASL/SCRAM or SASL/GSSAPI.
```

> Streaming data into or from Apache Kafka involves creating data bridges to Kafka in two roles: producer (sends messages to Kafka) and consumer (receives messages from Kafka). EMQX enables you to create data bridges in either of the roles.

Reviewer: ```suggestion
Streaming data into or from Apache Kafka involves creating data bridges to Kafka in two roles: producer (sends messages to Kafka) and consumer (receives messages from Kafka). EMQX enables you to create data bridges in either of the roles. This page introduces how to create the Kafka data bridge both as a producer and consumer role. 
```

> Streaming data into Apache Kafka involves creating data bridges to Kafka in two roles: producer (sends messages to Kafka) and consumer (receives messages from Kafuka). EMQX enables you to create data bridges in either or both of the roles.

Reviewer 2: Nit: consumer streams data _out of_ Kafka.  Also, each bridge can only be one of producer or consumer, but not both.

```suggestion
Streaming data into or from Apache Kafka involves creating data bridges to Kafka in two roles: producer (sends messages to Kafka) and consumer (receives messages from Kafka). EMQX enables you to create data bridges in either of the roles.
```

<!-- 根据情况编写，包含必须的前置知识点、软件版本要求、需要预先创建/初始化的操作。 -->
- Knowledge about EMQX data integration [rules](./rules.md)
- Knowledge about [data bridge](./data-bridges.md)

<!-- 列举功能或性能方面的亮点，如支持批处理、支持异步模式、双向数据桥接，链接到对应的功能介绍章节。 -->

:::

## Feature List

> - [Connection pool](./data-bridges.md#connection-pool) <!-- TODO 确认改版后知否支持-->

Reviewer 2: There's no configurable connection pool for Kafka:

```suggestion
```

   
   - **Message Value**: Kafka message value. Insert a string here, either a plain string or a string containing placeholders (${var}).
   
   - Advanced settings (optional): Set the **Max Batch Bytes**, **Compression**, and **Partition Strategy** as your business needs.
   
   :::
   
   ::: tab Configure as Consumer Role
   
   - Fill the required fields (marked with an asterisk). 
   
   - Input a name for the data bridge. The name should be a combination of upper/lower case letters and numbers.
   
   - Input the connection information. Input `127.0.0.1:9092` for the **Bootstrap Hosts**. For the other fields set as the actual condition.
   
   - The **Topic Mapping** field must contain at least one Kafka-to-MQTT topic mapping. The **MQTT Payload Template** subfield specifies the MQTT payload that should be used, and has the following Kafka message fields available for templating:
   
     | Field Name | Description                                                  |
     | ---------- | ------------------------------------------------------------ |
     | `headers`  | An object containing string key-value pairs                  |

>      | `key`      | Kafka message key (encoded by the chosen key encoding)       |

Reviewer: ```suggestion
     | `key`      | Kafka message key (uses the same encoding method as the selected key)       |
```

     | `offset`   | Offset for the message in Kafka's topic partition            |
     | `topic`    | Originating Kafka topic                                      |
     | `ts`       | Message timestamp                                            |
     | `ts_type`  | Message timestamp type, which is one of `create`, `append` or `undefined` |

>      | `value`    | Kafka message value (encoded by the chosen value encoding)   |

Reviewer: ```suggestion
     | `value`    | Kafka message value (uses the same encoding method as the selected key)   |
```

     
     ```json
     {
           "value": "value",
           "ts_type": "create",
           "ts": 1679665968238,
           "topic": "my-kafka-topic",
           "offset": 2,
           "key": "key",
           "headers": {"header_key": "header_value"}
          }
     ```
     
     Subfields from the Kafka message may be accessed with dot notation. For example, `${.value}` will resolve to the Kafka message value, and `${.headers.h1}` will resolve to the value of the `h1` Kafka header if such a subfield exists.  Absent values will be replaced by empty strings.
     
     **Note**: Each Kafka-to-MQTT topic mapping must contain a unique Kafka topic name.  That is, the Kafka topic must not be present in more than one mapping.
   
   :::
   
   ::::
   
5. Before clicking **Create**, you can click **Test Connection** to test that the bridge can connect to the Kafka server.

6. Click **Create**, you'll be offered the option of creating an associated rule. 

   - For the Kafka producer data bridge, click **Create Rule** to create an associated rule. For detailed operating steps, see [Create Rule for Kafka Producer Data Bridge](#create-rule-for-kafka-producer-data-bridge). 

>    - For the Kafka consumer data bridge, It's not strictly necessary to create a rule.

Reviewer 2: ```suggestion
   - For the Kafka consumer data bridge, it's not strictly necessary to create a rule.
```

bin/kafka-server-start.sh config/kraft/server.properties
```

For detailed operation steps, you may refer to the [Quick Start section in Kafka Documentation](https://kafka.apache.org/documentation/#quickstart).

### Create Kafka Topics

> Relevant Kafka topics should be created before creating the data bridge in EMQX. Use the commands below to create two topics in Kafka:  `testtopic-in` and `testtopic-out`. 

Reviewer: it would be great if we could also explain which topic is for which consumer

   
   - Input the connection information. Input `127.0.0.1:9092` for the **Bootstrap Hosts**. For the other fields set as the actual condition.
   
   - **Source MQTT Topic**: Set the MQTT topics to create the data bridge. In this example, it is set to `t/#`, indicating all MQTT messages matching this topic will be sent to Kafka. You can also leave it blank, and create a [rule](#create-rule-for-kafka-producer-data-bridge) to specify data to be sent to Kafka.
   
   - **Kafka Topic Name**: Input `testtopic-in` (the Kafka topic created before). Note: Variables are not supported here.
   
   - **Message Key**: Kafka message key. Insert a string here, either a plain string or a string containing placeholders (${var}).
   
   - **Message Value**: Kafka message value. Insert a string here, either a plain string or a string containing placeholders (${var}).
   
   - Advanced settings (optional): Set the **Max Batch Bytes**, **Compression**, and **Partition Strategy** as your business needs.
   
   :::
   
   ::: tab Configure as Consumer Role
   
   - Fill the required fields (marked with an asterisk). 
   

>    - Input a name for the data bridge. The name should be a combination of upper/lower case letters or numbers.

Reviewer: ```suggestion
   - Input a name for the data bridge. The name should be a combination of upper/lower case letters and numbers.
```

   
   - Input the connection information. Input `127.0.0.1:9092` for the **Bootstrap Hosts**. For the other fields set as the actual condition.
   
   - The **Topic Mapping** field must contain at least one Kafka-to-MQTT topic mapping. The **MQTT Payload Template** subfield specifies the MQTT payload that should be used, and has the following Kafka message fields available for templating:
   
     | Field Name | Description                                                  |
     | ---------- | ------------------------------------------------------------ |
     | `headers`  | An object containing string key-value pairs                  |

>      | `key`      | Kafka message key (encoded by the chosen key encoding)       |

Reviewer: encoded xxxx by encoding? 这句话不太理解

     | `offset`   | Offset for the message in Kafka's topic partition            |
     | `topic`    | Originating Kafka topic                                      |
     | `ts`       | Message timestamp                                            |
     | `ts_type`  | Message timestamp type, which is one of `create`, `append` or `undefined` |

>      | `value`    | Kafka message value (encoded by the chosen value encoding)   |

Reviewer: | `value`    | Kafka message value (encoded by the chosen value encoding)   |这句话没太看懂

   
     The default value for **MQTT Payload Template** is `${.}`, which includes all available data encoded as a JSON object.  For example, choosing `${.}` as a template will produce the following for a Kafka message:
     
     ```json
     {
           "value": "value",
           "ts_type": "create",
           "ts": 1679665968238,
           "topic": "my-kafka-topic",
           "offset": 2,
           "key": "key",
           "headers": {"header_key": "header_value"}
          }
     ```
     

>      Subfields from the Kafka message may be accessed with dot notation. For example: `${.value}` will resolve to the Kafka message value, and `${.headers.h1}` will resolve to the value of the `h1` Kafka header, if present.  Absent values will be replaced by empty strings.

Reviewer: 这里的 if present 是不是有点奇怪？

     
     **Note**: Each Kafka-to-MQTT topic mapping must contain a unique Kafka topic name.  That is, the Kafka topic must not be present in more than one mapping.
     
   
   :::
   
   ::::
   
5. Before clicking **Create**, you can click **Test Connection** to test that the bridge can connect to the Kafka server.

6. Click **Create**, you'll be offered the option of creating an associated rule. 

>    - For the Kafka producer data bridge, click **Create Rule** or go to [Create Rule for Kafka Producer Data Bridge](#create-rule-for-kafka-producer-data-bridge) to create an associated rule. 

Reviewer: ```suggestion
   - For the Kafka producer data bridge, click **Create Rule** to create an associated rule. For detailed operating steps, see [Create Rule for Kafka Producer Data Bridge](#create-rule-for-kafka-producer-data-bridge). 
```

:::

Now the Kafka data bridge should appear in the data bridge list (**Data Integration** -> **Data Bridge**) with **Resource Status** as **Connected**.

### Create Rule for Kafka Producer Data Bridge

1. Go to EMQX Dashboard, click **Data Integration** -> **Rules**.

2. Click **Create** on the top right corner of the page.

3. Input, for example, `my_rule` as the rule ID.

4.  Input the following statement in the **SQL Editor** if you want to save the MQTT messages under topic `t/#`  to Kafka. 

   Note: If you want to specify your own SQL syntax, make sure that you have included all fields required by the data bridge in the `SELECT` part.

```sql
SELECT
  *
FROM
  "t/#"
```

5. Click the **Add Action** button, select **Forwarding with Data Bridge** from the dropdown list and then select the data bridge you just created under **Data bridge**. Then click the **Add** button.

6. Click **Create** at the page bottom to finish the creation.

> Now we have successfully created the data bridge to Kafka producer data bridge. You can click **Data Integration** -> **Flows** to view the topology. It can be seen that the messages under topic `t/#`  are sent and saved to Kafka after parsing by rule  `my_rule`.

Reviewer: ```suggestion
Now you have successfully created the data bridge to Kafka producer data bridge. You can click **Data Integration** -> **Flows** to view the topology. It can be seen that the messages under topic `t/#`  are sent and saved to Kafka after parsing by rule  `my_rule`.
```
